src/replica/HttpAsyncClient.cc

```diff
@@ -0,0 +1,399 @@
+    _assertState(lock, context, {State::CREATED});
+    try {
+        _parsUrl(lock);
+        _prepareRequest(lock);
+        _resolve(lock);
+        if (_expirationIvalSec) {
+            _expirationTimer.expires_from_now(boost::posix_time::seconds(_expirationIvalSec));
+            _expirationTimer.async_wait(bind(&HttpAsyncClient::_expired, shared_from_this(), _1));
+        }
+        _state = State::IN_PROGRESS;
+    } catch(exception const& ex) {
+        _finish(lock, State::FAILED, ex.what());
+        throw;
+    }
+}
+
+
+void HttpAsyncClient::cancel() {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    switch (_state) {
+        case State::CREATED:
+            _finish(lock, State::CANCELLED);
+            return;
+        case State::IN_PROGRESS:
+            _finish(lock, State::CANCELLED);
+            return;
+        default:
+            // Nothing to be done in the final states of the request.
+            break;
+    }
+}
+
+
+string HttpAsyncClient::error() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    return _error;
+}
+
+
+int HttpAsyncClient::responseCode() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::FINISHED, State::BODY_LIMIT_ERROR});
+    auto const& header = _res.get().base();
+    return header.result_int();
+}
+
+
+map<string, string> const& HttpAsyncClient::responseHeader() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::FINISHED, State::BODY_LIMIT_ERROR});
+    return _responseHeader;
+}
+
+
+string const& HttpAsyncClient::responseBody() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::FINISHED});
+    return _res.get().body();
+}
+
+
+size_t HttpAsyncClient::responseBodySize() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::FINISHED});
+    return _res.get().body().size();
+}
+
+
+void HttpAsyncClient::_parsUrl(util::Lock const& lock) {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    regex re("^[ ]*http://([^:^/]+)?(:([0-9]+))?(/([^ ]*))?[ ]*$", regex::extended);
+    smatch match;
+    if (!regex_search(_url, match, re)) {
+        throw invalid_argument(context + "incorrect syntax of the URL '" + _url + "'.");
+    }
+    if (match.size() != 6) {
+        throw invalid_argument(context + "problem with the regular expression when parsing URL '" + _url + "'.");
+    }
+    auto const parseVal = [](string const& in, string& out) {
+        if (!in.empty()) out = in;
+    };
```

Comment:
This is not the lambda that I can approve, even if it's short.

Reply:
What's wrong with this lambda? Look at the next 3 lines of code where it's used. A value of the `in` parameter is extracted from the result of matching a regular expression. Using the lambda here simplifies the value analysis and prevets dereferencing the input value from the above mentioned result twice.

Reply:
Couple of points:
- I don't see this as an improvement, it makes code longer but not clearer
- the name is confusing, `parseVal` does not parse anything, just assigns something to something
- if you care about performance you should probably write something like `if (match[1].matched) _host = match[1];` to avoid creating strings to just check that they are empty.